overlord/configstate/configcore/snapshots.go

```diff
@@ -0,0 +1,54 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/overlord/configstate/config"
+)
+
+func init() {
+	// add supported configuration of this module
+	supportedConfigurations["core.automatic-snapshots.expiration"] = true
+}
+
+func validateAutomaticSnapshotsExpiration(tr config.Conf) error {
+	expirationStr, err := coreCfg(tr, "automatic-snapshots.expiration")
+	if err != nil {
+		return err
+	}
+	if expirationStr != "" {
+		dur, err := time.ParseDuration(expirationStr)
+		if err != nil {
+			return fmt.Errorf("automatic-snapshots.expiration cannot be parsed: %v", err)
+		}
+		if dur > 0 && dur < time.Hour*24 {
+			return fmt.Errorf("automatic-snapshots.expiration must be 0 to disable automatic snapshots, or a value greater than 24 hours")
+		}
+		// special-case "0" (with no unit): it's a valid duration (any other number with unit omitted isn't), but when left as is it would
+		// be stored as int64 instead of a string representing duration, causing issue when reading from the state.
+		if expirationStr == "0" {
+			tr.Set("core", "automatic-snapshots.expiration", "0s")
```

Comment:
This slightly unfortunate and ugly. An alternative would be to - instead of storing duration as a string - simply take parsed int64 duration and always store it the config in place of the value coming from `snap set` (with no special case needed). The downside is that `snap get ...` will return a cryptic duration int (a large one), instead of something like "24h"...

Reply:
the other approach would be to support keeping there both  a duration string,  or a positive number of minutes, or use a special string instead of 0  but that's also why I would like to turn expiration into it's positive opposite above, because it would make finding such a string easier,  for example automatic-snapshots.expiration=no  would mean the opposite.  maybe again @chipaca or @degville have ideas

Reply:
I don't think you should be able to set a the retention to less than a minimum. What that minimum is depends on whether we can change it for tests or not :-) but, in real life, retention of less than an hour seems unlikely, less than a minute is pointless.

Reply:
Yes, absolutely, retention of 24h is the minimum. "0s" was a special case for disabling automatic snapshots, now changed to "no" per above comments.